### PR TITLE
Introduce RemoteTokenTracker in BlobStore to clean up permanent delete tombstone

### PR DESCRIPTION
**Pull request by jsjtzyy**
Introduce RemoteTokenTracker to help keep track of replication tokens from all peer replicas. The tokens maintained within this tracker will be consumed by compaction.

### CODE REVIEW

File: ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.replication.FindToken;
+import java.util.HashMap;
+import java.util.List;
```
Comment (SophieGuo410) [inline ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java:19]: minor: not used.

File: ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  public RemoteTokenTracker(ReplicaId localReplica) {
+    this.localReplica = localReplica;
+    localReplica.getPeerReplicaIds().forEach(r -> {
+      String hostnameAndPath = r.getDataNodeId().getHostname() + DELIMITER + r.getReplicaPath();
+      peerReplicaAndToken.put(hostnameAndPath, new StoreFindToken());
+    });
+  }
+
+  void updateTokenFromPeerReplica(FindToken token, String remoteHostName, String remoteReplicaPath) {
```
Comment (SophieGuo410) [inline ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java:44]: Probably want to add design doc for this method.

File: ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * {@link RemoteTokenTracker} tracks tokens from all peer replicas and updates them when handling metadata request from
+ * peer node.
+ */
+public class RemoteTokenTracker {
+  private static final String DELIMITER = ":";
+  // The key of peerReplicaAndToken is a string containing hostname and path of peer replica.
+  // For example: localhost:/mnt/u001/p1
+  private ConcurrentMap<String, FindToken> peerReplicaAndToken = new ConcurrentHashMap<>();
+  private ReplicaId localReplica;
```
Comment (SophieGuo410) [inline ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java:34]: could be final?

**Review by SophieGuo410 — COMMENTED** [review]
Overall LGTM!

File: ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java (hunk 3 above)
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java:34]: yep, will fix this.

File: ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java (hunk 2 above)
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java:44]: sure, updated

File: ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java (hunk 1 above)
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java:19]: removed

**Review by SophieGuo410 — APPROVED** [review]
lgtm!

**Review by ankagrawal — APPROVED** [review]
lgtm!

**Review by justinlin-linkedin — COMMENTED** [review]
would you please also rebase from master branch and then update this PR?

**Comment by jsjtzyy** [comment]
> would you please also rebase from master branch and then update this PR?

sure, sorry I didn't notice the conflict, will fix it shortly

File: ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * {@link RemoteTokenTracker} tracks tokens from all peer replicas and updates them when handling metadata request from
+ * peer node.
+ */
+public class RemoteTokenTracker {
+  private static final String DELIMITER = ":";
+  private final ReplicaId localReplica;
+  // The key of peerReplicaAndToken is a string containing hostname and path of peer replica.
+  // For example: localhost:/mnt/u001/p1
+  private ConcurrentMap<String, FindToken> peerReplicaAndToken = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java:33]: I partially agree. The `ReplicaId` needs to reconstructed on source node and replication metadata request contains remote hostname/replica path. There might be an edge case where remote replica is removed while source node hasn't noticed the change, which means it's not able to construct replicaId. In the context of remote token tracker, a pair of hostname and replica path should be sufficient to uniquely identify a remote peer replica.

**Review by justinlin-linkedin — APPROVED** [review]

**Pull request merged**